In [1]:
import pandas as pd
df = pd.read_csv("heart.csv")
df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

🟢 Removing outliers using zscore

In [2]:
df["resting_bp_zscore"] = (df.RestingBP - df.RestingBP.mean())/ df.RestingBP.std()

In [3]:
df[df["resting_bp_zscore"]>3]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  
109    106              N      0.0       Up             0           3.111322  
241    142              Y      2.0     Flat             1           3.651449  
365    140              Y      1.0     Flat             1           3.651449  
399     70              N      0.0     Flat             1           3.651449  
592    150              Y      2.0     Down             1           3.111322  
732    133              Y      4.0     Down             1           3.651449  
759    195              N      0.0       Up             1           3.219347

In [4]:
df[df['resting_bp_zscore']<-3]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
449   55   M           NAP          0            0          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  
449    155              N      1.5     Flat             1          -7.151097

In [5]:
df_new = df[(df.resting_bp_zscore>-3)&(df.resting_bp_zscore<3)]

In [6]:
df_new

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  
0      172              N      0.0       Up             0           0.410685  
1      156              N      1.0     Flat             1           1.490940  
2       98              N      0.0       Up             0          -0.129442  
3      108              Y      1.5     Flat             1           0.302660  
4      122              N      0.0       Up             0           0.950812  
..     ...            ...      ...      ...           ...                ...  
913    132              N      1.2     Flat             1          -1.209697  
914    141              N      3.4     Flat             1           0.626736  
915    115              Y      1.2     Flat             1          -0.129442  
916    174              N      0.0     Flat             1          -0.129442  
917    173              N      0.0       Up             0           0.302660  

[910 rows x 13 columns]

In [7]:
df_new['Cholesterol_zscore'] = (df_new.Cholesterol - df_new.Cholesterol.mean())/df_new.Cholesterol.std()

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\2160507423.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['Cholesterol_zscore'] = (df_new.Cholesterol - df_new.Cholesterol.mean())/df_new.Cholesterol.std()


In [8]:
df_new

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  
0              0.823475  
1             -0.175198  
2              0.768502  
3              0.136315  
4             -0.037766  
..                  ...  
913            0.594422  
914           -0.056090  
915           -0.624143  
916            0.337882  
917           -0.221009  

[910 rows x 14 columns]

In [9]:
df_new1 = df_new[(df_new.Cholesterol_zscore>-3)&(df_new.Cholesterol_zscore<3)]

In [10]:
df_new1

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  
0              0.823475  
1             -0.175198  
2              0.768502  
3              0.136315  
4             -0.037766  
..                  ...  
913            0.594422  
914           -0.056090  
915           -0.624143  
916            0.337882  
917           -0.221009  

[907 rows x 14 columns]

In [11]:
df_new1['MaxHR_zscore'] = (df_new1.MaxHR - df_new1.MaxHR.mean())/df_new1.MaxHR.std()

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\2463019731.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new1['MaxHR_zscore'] = (df_new1.MaxHR - df_new1.MaxHR.mean())/df_new1.MaxHR.std()


In [12]:
df_new1

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  MaxHR_zscore  
0              0.823475      1.385705  
1             -0.175198      0.755693  
2              0.768502     -1.528100  
3              0.136315     -1.134343  
4             -0.037766     -0.583082  
..                  ...           ...  
913            0.594422     -0.189325  
914           -0.056090      0.165057  
915           -0.624143     -0.858713  
916            0.337882      1.464457  
917           -0.221009      1.425081  

[907 rows x 15 columns]

In [13]:
df_new2 = df_new1[(df_new1.MaxHR_zscore>-3)&(df_new1.MaxHR_zscore<3)]

In [14]:
df_new2

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  MaxHR_zscore  
0              0.823475      1.385705  
1             -0.175198      0.755693  
2              0.768502     -1.528100  
3              0.136315     -1.134343  
4             -0.037766     -0.583082  
..                  ...           ...  
913            0.594422     -0.189325  
914           -0.056090      0.165057  
915           -0.624143     -0.858713  
916            0.337882      1.464457  
917           -0.221009      1.425081  

[906 rows x 15 columns]

In [15]:
df_new2['Oldpeak_zscore'] = (df_new2.Oldpeak - df_new2.Oldpeak.mean()) / df_new2.Oldpeak.std()

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\557192073.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new2['Oldpeak_zscore'] = (df_new2.Oldpeak - df_new2.Oldpeak.mean()) / df_new2.Oldpeak.std()


In [16]:
df_new2

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  MaxHR_zscore  Oldpeak_zscore  
0              0.823475      1.385705       -0.830801  
1             -0.175198      0.755693        0.108320  
2              0.768502     -1.528100       -0.830801  
3              0.136315     -1.134343        0.577881  
4             -0.037766     -0.583082       -0.830801  
..                  ...           ...             ...  
913            0.594422     -0.189325        0.296145  
914           -0.056090      0.165057        2.362211  
915           -0.624143     -0.858713        0.296145  
916            0.337882      1.464457       -0.830801  
917           -0.221009      1.425081       -0.830801  

[906 rows x 16 columns]

In [17]:
final_df = df_new2[(df_new2.Oldpeak_zscore>-3)&(df_new2.Oldpeak_zscore<3)]

In [18]:
final_df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  resting_bp_zscore  \
0      172              N      0.0       Up             0           0.410685   
1      156              N      1.0     Flat             1           1.490940   
2       98              N      0.0       Up             0          -0.129442   
3      108              Y      1.5     Flat             1           0.302660   
4      122              N      0.0       Up             0           0.950812   
..     ...            ...      ...      ...           ...                ...   
913    132              N      1.2     Flat             1          -1.209697   
914    141              N      3.4     Flat             1           0.626736   
915    115              Y      1.2     Flat             1          -0.129442   
916    174              N      0.0     Flat             1          -0.129442   
917    173              N      0.0       Up             0           0.302660   

     Cholesterol_zscore  MaxHR_zscore  Oldpeak_zscore  
0              0.823475      1.385705       -0.830801  
1             -0.175198      0.755693        0.108320  
2              0.768502     -1.528100       -0.830801  
3              0.136315     -1.134343        0.577881  
4             -0.037766     -0.583082       -0.830801  
..                  ...           ...             ...  
913            0.594422     -0.189325        0.296145  
914           -0.056090      0.165057        2.362211  
915           -0.624143     -0.858713        0.296145  
916            0.337882      1.464457       -0.830801  
917           -0.221009      1.425081       -0.830801  

[899 rows x 16 columns]

🟢 After clearing the outliers here we can see clearly the dataset has reduced from 918 to 899

🟢 And We are removing the columns which are not needed for data modeling

In [19]:
final_df.drop(["resting_bp_zscore","Cholesterol_zscore","MaxHR_zscore","Oldpeak_zscore"],axis=1,inplace=True)

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\4095995727.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df.drop(["resting_bp_zscore","Cholesterol_zscore","MaxHR_zscore","Oldpeak_zscore"],axis=1,inplace=True)


Here we got the dataframe we all the columns we need

In [20]:
final_df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[899 rows x 12 columns]

🟢 Converting the text columns with one hot encoding

In [21]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
final_df['Sex'] = label_encoder.fit_transform(final_df['Sex'])

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\345462484.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['Sex'] = label_encoder.fit_transform(final_df['Sex'])


In [22]:
final_df

Age  Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40    1           ATA        140          289          0     Normal   
1     49    0           NAP        160          180          0     Normal   
2     37    1           ATA        130          283          0         ST   
3     48    0           ASY        138          214          0     Normal   
4     54    1           NAP        150          195          0     Normal   
..   ...  ...           ...        ...          ...        ...        ...   
913   45    1            TA        110          264          0     Normal   
914   68    1           ASY        144          193          1     Normal   
915   57    1           ASY        130          131          0     Normal   
916   57    0           ATA        130          236          0        LVH   
917   38    1           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[899 rows x 12 columns]

In [23]:
final_df['ChestPainType'] = label_encoder.fit_transform(final_df['ChestPainType'])
final_df['RestingECG'] = label_encoder.fit_transform(final_df['RestingECG'])
final_df['ExerciseAngina'] = label_encoder.fit_transform(final_df['ExerciseAngina'])
final_df['ST_Slope'] = label_encoder.fit_transform(final_df['ST_Slope'])

C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\2286842735.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['ChestPainType'] = label_encoder.fit_transform(final_df['ChestPainType'])
C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\2286842735.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['RestingECG'] = label_encoder.fit_transform(final_df['RestingECG'])
C:\Users\jeyar\AppData\Local\Temp\ipykernel_6360\2286842735.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy

In [24]:
final_df

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    1              1        140          289          0           1   
1     49    0              2        160          180          0           1   
2     37    1              1        130          283          0           2   
3     48    0              0        138          214          0           1   
4     54    1              2        150          195          0           1   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    1              3        110          264          0           1   
914   68    1              0        144          193          1           1   
915   57    1              0        130          131          0           1   
916   57    0              1        130          236          0           0   
917   38    1              2        138          175          0           1   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0      172               0      0.0         2             0  
1      156               0      1.0         1             1  
2       98               0      0.0         2             0  
3      108               1      1.5         1             1  
4      122               0      0.0         2             0  
..     ...             ...      ...       ...           ...  
913    132               0      1.2         1             1  
914    141               0      3.4         1             1  
915    115               1      1.2         1             1  
916    174               0      0.0         1             1  
917    173               0      0.0         2             0  

[899 rows x 12 columns]

🟢 Here you can see the dataset after onehot encoding process

In [25]:
final_df.describe()

Age         Sex  ChestPainType   RestingBP  Cholesterol  \
count  899.000000  899.000000     899.000000  899.000000   899.000000   
mean    53.497219    0.789766       0.785317  132.027809   198.005562   
std      9.456073    0.407701       0.956496   17.120895   107.157779   
min     28.000000    0.000000       0.000000   80.000000     0.000000   
25%     47.000000    1.000000       0.000000  120.000000   174.500000   
50%     54.000000    1.000000       0.000000  130.000000   222.000000   
75%     60.000000    1.000000       2.000000  140.000000   266.000000   
max     77.000000    1.000000       3.000000  185.000000   518.000000   

        FastingBS  RestingECG       MaxHR  ExerciseAngina     Oldpeak  \
count  899.000000  899.000000  899.000000      899.000000  899.000000   
mean     0.232481    0.991101  136.917686        0.403782    0.861513   
std      0.422649    0.629569   25.356740        0.490928    1.007626   
min      0.000000    0.000000   63.000000        0.000000   -2.000000   
25%      0.000000    1.000000  120.000000        0.000000    0.000000   
50%      0.000000    1.000000  138.000000        0.000000    0.500000   
75%      0.000000    1.000000  156.000000        1.000000    1.500000   
max      1.000000    2.000000  202.000000        1.000000    4.000000   

         ST_Slope  HeartDisease  
count  899.000000    899.000000  
mean     1.373749      0.547275  
std      0.601057      0.498037  
min      0.000000      0.000000  
25%      1.000000      0.000000  
50%      1.000000      1.000000  
75%      2.000000      1.000000  
max      2.000000      1.000000

🟢 Applying scaling to the model before training the model

In [26]:
X = final_df.drop("HeartDisease",axis=1)
y = final_df['HeartDisease']

In [27]:
X,y

(     Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
 0     40    1              1        140          289          0           1   
 1     49    0              2        160          180          0           1   
 2     37    1              1        130          283          0           2   
 3     48    0              0        138          214          0           1   
 4     54    1              2        150          195          0           1   
 ..   ...  ...            ...        ...          ...        ...         ...   
 913   45    1              3        110          264          0           1   
 914   68    1              0        144          193          1           1   
 915   57    1              0        130          131          0           1   
 916   57    0              1        130          236          0           0   
 917   38    1              2        138          175          0           1   
 
      MaxHR  ExerciseAngina  Oldpeak  

🟢 Here we are using StandardScaler from sklearn for getting the standard deviation of all the features

In [28]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[-1.42815446,  0.515943  ,  0.2245723 , ..., -0.8229452 ,
        -0.85546862,  1.04249607],
       [-0.47585532, -1.93819859,  1.27063705, ..., -0.8229452 ,
         0.13751561, -0.62216462],
       [-1.7455875 ,  0.515943  ,  0.2245723 , ..., -0.8229452 ,
        -0.85546862,  1.04249607],
       ...,
       [ 0.3706328 ,  0.515943  , -0.82149245, ...,  1.21514774,
         0.33611246, -0.62216462],
       [ 0.3706328 , -1.93819859,  0.2245723 , ..., -0.8229452 ,
        -0.85546862, -0.62216462],
       [-1.63977649,  0.515943  ,  1.27063705, ..., -0.8229452 ,
        -0.85546862,  1.04249607]])

🟢 Here we are going to build a classification model , Here are chosing the Logistic Regression model for constructing the model

In [29]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X_scaled,y,test_size=0.2,random_state=30)

In [34]:
from sklearn.linear_model import LogisticRegression

logistic_regression = LogisticRegression()
logistic_regression.fit(X_train,y_train)
logistic_regression.score(X_test,y_test)

0.8333333333333334

🟢 Saving the model as the model has the best performance before pca 

🟢 The Model performance after PCA was worse so i am using this model for predictions

In [35]:
import pickle
filename = 'final_model.sav'
saved_model = pickle.dump(logistic_regression,open(filename,'wb'))

In [36]:
import pickle

loaded_model = pickle.load(open('final_model.sav','rb'))
result = loaded_model.score(X_test,y_test)
result

0.8333333333333334

In [37]:
y_pred = loaded_model.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(loaded_model.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.83


In [38]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.86      0.83        83
           1       0.87      0.81      0.84        97

    accuracy                           0.83       180
   macro avg       0.83      0.83      0.83       180
weighted avg       0.84      0.83      0.83       180



In [39]:
print(X_test[7])
print(y_pred[7])

[ 1.11130991  0.515943   -0.82149245 -0.70291322  0.44813503 -0.5503622
 -1.57512958 -1.61457911  1.21514774  1.3290967  -2.2868253 ]
1


🟢 Here we are using PCA for dimension reduction

In [40]:
from sklearn.decomposition import PCA

pca = PCA(0.95)

X_pca=pca.fit_transform(X)
X_pca.shape

(899, 2)

In [41]:
pca.explained_variance_ratio_

array([0.92109746, 0.05064983])

In [42]:
pca.n_components_

2

In [43]:
X_train_pca,X_test_pca,y_train,y_test = train_test_split(X_pca,y,test_size=0.2,random_state=30)

In [44]:
model2 = LogisticRegression()
model2.fit(X_train_pca,y_train)
model2.score(X_test_pca,y_test)

0.6666666666666666

🟢 Here you can see the model performance is reduced while we are applying PCA to the dataset

🟢 Accuracy of the model without PCA:83
🟢 Accuracy of the model with PCA:66

🟢 So using PCA in this dataset is not recommended